File: ios/Jazzify/EarTraining/Engine/EarTrainingPracticeSpeed.swift
import Foundation

enum EarTrainingPracticeSpeed {
    static let practiceSpeedMinPercent = 40
    static let practiceSpeedMaxPercent = 100

    static func clampPracticeSpeedPercent(_ percent: Int) -> Int {
        max(practiceSpeedMinPercent, min(practiceSpeedMaxPercent, percent))
    }

    static func practiceSpeedRatio(_ percent: Int) -> Double {
        Double(clampPracticeSpeedPercent(percent)) / 100.0
    }

    static func scalePracticeTargetTimeSec(_ targetTimeSec: Double, speedPercent: Int) -> Double {
        let ratio = practiceSpeedRatio(speedPercent)
        guard ratio < 1 else { return targetTimeSec }
        return targetTimeSec / ratio
    }

    static func scalePracticeTimingWindowSec(_ windowSec: Double, speedPercent: Int) -> Double {
        let ratio = practiceSpeedRatio(speedPercent)
        guard ratio < 1 else { return windowSec }
        return windowSec / ratio
    }

    static func effectivePracticeBpm(_ bpm: Int, speedPercent: Int) -> Int {
        max(1, Int((Double(bpm) * practiceSpeedRatio(speedPercent)).rounded()))
    }

    static func scalePracticePhraseLoopEndSec(_ loopEndSec: Double, speedPercent: Int) -> Double {
        let ratio = practiceSpeedRatio(speedPercent)
        guard ratio < 1 else { return loopEndSec }
        return loopEndSec / ratio
    }

    /// 練習タイムライン秒を、未ストレッチ PCM / ファイル上の秒へ変換する（`AVAudioUnitTimePitch` 再生用）。
    static func practiceBufferOffsetSec(timelineOffsetSec: Double, speedPercent: Int) -> Double {
        max(0, timelineOffsetSec) * practiceSpeedRatio(speedPercent)
    }
}

File: src/utils/earTrainingPracticeSpeed.ts
/** 練習速度の下限（%） */
export const PRACTICE_SPEED_MIN_PERCENT = 40;

/** 練習速度の上限（%）・デフォルト */
export const PRACTICE_SPEED_MAX_PERCENT = 100;

export const clampPracticeSpeedPercent = (percent: number): number =>
  Math.max(
    PRACTICE_SPEED_MIN_PERCENT,
    Math.min(PRACTICE_SPEED_MAX_PERCENT, Math.trunc(percent)),
  );

export const practiceSpeedRatio = (percent: number): number =>
  clampPracticeSpeedPercent(percent) / 100;

/** 練習速度を反映した判定・スクロール用ターゲット時刻（秒） */
export const scalePracticeTargetTimeSec = (
  targetTimeSec: number,
  speedPercent: number,
): number => {
  const ratio = practiceSpeedRatio(speedPercent);
  if (ratio >= 1) {
    return targetTimeSec;
  }
  return targetTimeSec / ratio;
};

/** カウントイン等に使う effective BPM */
export const effectivePracticeBpm = (bpm: number, speedPercent: number): number =>
  bpm * practiceSpeedRatio(speedPercent);

/** 練習速度を反映した判定窓・ヒント段階などのタイミング幅（秒） */
export const scalePracticeTimingWindowSec = (
  windowSec: number,
  speedPercent: number,
): number => {
  const ratio = practiceSpeedRatio(speedPercent);
  if (ratio >= 1) {
    return windowSec;
  }
  return windowSec / ratio;
};

/** フレーズループ終了秒を練習速度に合わせてスケール */
export const scalePracticePhraseLoopEndSec = (
  loopEndSec: number,
  speedPercent: number,
): number => {
  const ratio = practiceSpeedRatio(speedPercent);
  if (ratio >= 1) {
    return loopEndSec;
  }
  return loopEndSec / ratio;
};

export const formatPracticeSpeedPercentLabel = (percent: number): string =>
  `${clampPracticeSpeedPercent(percent)}%`;
